Resolves #53

thedanvail opened the pull request:
Adds a more graceful exit when the directory does not exist.

If the desired behavior is any different (for example, printing the absolute path of the nonexistent directory), I can update with the absolute path.

aviramha reviewed (changes requested):
Thanks for your contribution!
I have left few notes in the code.
Feel free to ask for clarifications/guidance if needed.
Also, you are more than welcome to consult with us in our Discord server.

thedanvail commented:
@aviramha I tried to take all of your comments into consideration as I updated using `anyhow` to handle an invalid path. 
Additional meta-information:
1. This is my first open source PR, so I apologize if I'm not quite following best practices
2. I found this issue through the `This Week In Rust` call for collaboration - just so you know!

aviramha commented:
@thedanvail 
1. Great! Thanks for choosing us to be your first contribution. Beware, OSS contributing can be addictive!
2. Cool! Thanks for letting us know :)

aviramha commented on mirrord-cli/src/main.rs:
```diff
@@ -117,7 +120,7 @@ fn exec(args: &ExecArgs) -> Result<()> {
         std::env::set_var("MIRRORD_ACCEPT_INVALID_CERTIFICATES", "true");
     }
-    let library_path = extract_library(None);
+    let library_path = extract_library(None).unwrap();
```
you can use `?` here instead of unwrap

aviramha commented on mirrord-cli/src/main.rs:
```diff
@@ -138,7 +141,7 @@ fn main() {
         Commands::Exec(args) => exec(&args).unwrap(),
         Commands::Extract { path } => {
-            extract_library(Some(path));
+            extract_library(Some(path)).unwrap();
```
Change fn main to return `Result<()>` then you can use `?` here

thedanvail replied:
If I change the main function signature to return `Result<()>`, then we'd have to remove the `unwrap()`s/`?` to return the `Result` - or we can leave in the `unwrap()`s/`?` and leave out `Result` from the return type - which behavior would be preferred?

aviramha replied:
`?` is a syntatic sugar that returns when a `Result`/`Option` has `Err`/`None` else it returns the inner value.
So the change would to add `?` and change the return type to `Result`
https://doc.rust-lang.org/rust-by-example/std/result/question_mark.html

thedanvail replied:
I understand that, but the issue is that, if we leave in the `Result` return type, running cargo check gives me the following:

<a href="https://ibb.co/xFV8zKR"><img src="https://i.ibb.co/Sfq7VD9/Screen-Shot-2022-06-03-at-2-27-25-PM.png" alt="Screen-Shot-2022-06-03-at-2-27-25-PM" border="0"></a>

So the types don't match up if we add `Result` alongside unwrapping. I just want to make sure I'm making the changes you want - I apologize if I'm being difficult!

thedanvail replied:
I understand that, but the issue is, if I run `cargo check` with the following code

```
fn main() -> Result<()> {
    registry()
        .with(fmt::layer())
        .with(EnvFilter::from_default_env())
        .init();
    prompt_outdated_version();

    let cli = Cli::parse();
    match cli.commands {
        Commands::Exec(args) => exec(&args)?,
        Commands::Extract { path } => {
            extract_library(Some(path))?;
        }
    }
}
```

I receive the following error:

<a href="https://ibb.co/kHv9Y4L"><img src="https://i.ibb.co/p05RCfH/Screen-Shot-2022-06-03-at-2-44-03-PM.png" alt="Screen-Shot-2022-06-03-at-2-44-03-PM" border="0"></a>

aviramha replied:
Sorry but the image isn't available.. you can upload directly to GitHub by pasting the image here.
Considering the snippet, I think I know what the issue is. You can just drop the `?` in both arms and it'd work because both `exec` and `extract_library` return `Result`, so no need to unwrap it and you can just return their return value.

thedanvail replied:
Added in your changes - since `extract_library` returns a `Result<String>` instead of `Result<()>`, I just return a normal `Ok(())`. I think that should cover everything!

aviramha reviewed (changes requested):
The context can be great, but a few more tweaks are needed to merge this.
Thanks!

aviramha commented on CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ Previous versions had CHANGELOG per component, we decided to combine all reposit
 
 ## [Unreleased]
+### Changed
+- Added a graceful exit for system paths that do not exist.
```
Reword - "Added graceful exit for library extraction logic in case of error"

aviramha commented on mirrord-cli/src/main.rs:
```diff
@@ -60,19 +60,22 @@ const INJECTION_ENV_VAR: &str = "LD_PRELOAD";
         None => temp_dir().as_path().join(file_name),
     };
-    let mut file = File::create(file_path.clone()).unwrap();
+    let mut file = File::create(&file_path)
```
We can't know for sure the reason it failed is due to path not existing, so I'd just add using context the file path ("Path .. creation failed")

aviramha commented on mirrord-cli/src/main.rs:
```diff
@@ -136,9 +139,10 @@ fn main() {
-        Commands::Exec(args) => exec(&args).unwrap(),
+        Commands::Exec(args) => exec(&args),
         Commands::Extract { path } => {
             extract_library(Some(path));
```
I think you would have warning here if you wont add ?

thedanvail replied:
Nay - the function returns `Result<()>`, and because we added that to the `main() -> Result<()>` as well, this compiles as is. If I add the `?` operator - `Commands::Exec(args) => exec(&args)?,`, `cargo check` returns the following:

```
error[E0308]: `?` operator has incompatible types
   --> mirrord-cli/src/main.rs:142:33
    |
142 |         Commands::Exec(args) => exec(&args)?,
    |                                 ^^^^^^^^^^^^ expected enum `Result`, found `()`
    |
    = note: `?` operator cannot convert from `()` to `Result<(), anyhow::Error>`
    = note:   expected enum `Result<(), anyhow::Error>`
            found unit type `()`
help: try removing this `?`
    |
142 -         Commands::Exec(args) => exec(&args)?,
142 +         Commands::Exec(args) => exec(&args),
    | 
help: try wrapping the expression in `Ok`
    |
142 |         Commands::Exec(args) => Ok(exec(&args)?),
    |                                 +++            +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `mirrord` due to previous error
```

aviramha replied:
I think it'd be best to have `?` in the end of all arm branches (on exec + Extract) then outside the match branch (in the end of function) just to have `Ok(())` which means that if it got there, we're all good.

aviramha reviewed (changes requested):
Few small tweaks left :)
Sorry for being hard, but we want to make sure everything is great.
Also please fix conflicts with the main branch so we GH can run our automated tests.

thedanvail commented:
> Few small tweaks left :) Sorry for being hard, but we want to make sure everything is great. Also please fix conflicts with the main branch so we GH can run our automated tests.

No, please be as hard as you'd like! I'm trying to take as many notes as I can to get better at Rust and so I can be better at my next OSS PR. I sincerely appreciate your patience with me during this process!

I believe I have updated to reflect your suggestions and updated to fix the merge conflicts. I'm currently between a couple flights, so I didn't have time to check the status before the plane took off, but should be good now!

aviramha commented:
> > Few small tweaks left :) Sorry for being hard, but we want to make sure everything is great. Also please fix conflicts with the main branch so we GH can run our automated tests.
> 
> No, please be as hard as you'd like! I'm trying to take as many notes as I can to get better at Rust and so I can be better at my next OSS PR. I sincerely appreciate your patience with me during this process!
> 
> I believe I have updated to reflect your suggestions and updated to fix the merge conflicts. I'm currently between a couple flights, so I didn't have time to check the status before the plane took off, but should be good now!

Awesome, just see my last comment in the conversation about the `?`.

thedanvail commented:
Done and done!

thedanvail commented:
(Just updated with cargo fmt)

thedanvail commented:
I looked at the last CI run, and it looks like the `cargo clippy` issues are in `mirrord-layer`, which I did not touch in this PR. Let me know if I should do anything further.

aviramha commented:
Yes you're right. We're not binding clippy to a specific version so such cases can happen :)
Thanks for the contribution!
